src/perf.rs

```diff
@@ -0,0 +1,187 @@
+
+                // Don't currently support performance counter use on non-linux targets.
+                // `perfcnt` doesn't feature-gate `linux` so explicitly provide None's here rather
+                // than calling possibly wildly incorrect syscalls on non-linux targets.
+                Ok(PerfCounterCollection {
+                    precision,
+                    cpu_cycles: None,
+                    instructions_retired: None,
+                    cache_accesses: None,
+                    cache_misses: None,
+                })
+            }
+        }
+    }
+
+    fn start(&mut self) -> PerfCounterMeasure {
+        let time_start = self.precision.now();
+        if let Some(cpu_cycle_counter) = self.cpu_cycles.as_mut() {
+            cpu_cycle_counter
+                .reset()
+                .expect("if a counter could be created, it can be reset");
+        }
+        if let Some(instructions_retired) = self.instructions_retired.as_mut() {
+            instructions_retired
+                .reset()
+                .expect("if a counter could be created, it can be reset");
+        }
+        if let Some(cache_accesses) = self.cache_accesses.as_mut() {
+            cache_accesses
+                .reset()
+                .expect("if a counter could be created, it can be reset");
+        }
+        if let Some(cache_misses) = self.cache_misses.as_mut() {
+            cache_misses
+                .reset()
+                .expect("if a counter could be created, it can be reset");
+        }
+        PerfCounterMeasure {
+            time_start,
+            counters: self,
+        }
+    }
+
+    pub fn sample<F: FnOnce()>(&mut self, f: F) -> Sample {
+        let measurement = self.start();
+        f();
+        measurement.finish()
+    }
+
+    pub fn now(&self) -> Timestamp {
+        self.precision.now()
+    }
+}
+
+struct PerfCounterMeasure<'a> {
+    time_start: Timestamp,
+    counters: &'a mut PerfCounterCollection,
+}
+
+impl<'a> PerfCounterMeasure<'a> {
+    fn finish(self) -> Sample {
+        let time_end = self.counters.precision.now();
+        Sample {
+            clock_time: time_end - self.time_start,
+            cpu_cycles: self
+                .counters
+                .cpu_cycles
+                .as_mut()
+                .and_then(|c| c.read().ok())
+                .unwrap_or(0),
+            instructions_retired: self
+                .counters
+                .instructions_retired
+                .as_mut()
+                .and_then(|c| c.read().ok())
+                .unwrap_or(0),
+            cache_accesses: self
+                .counters
+                .cache_accesses
+                .as_mut()
+                .and_then(|c| c.read().ok())
+                .unwrap_or(0),
+            cache_misses: self
+                .counters
+                .cache_misses
+                .as_mut()
+                .and_then(|c| c.read().ok())
+                .unwrap_or(0),
+        }
```

Comment:
Is there no good way to capture the state of all of the counters at the same time? Hopefully the stuff in between each `c.read()` isn't significant but it would be nice...

Reply:
it's.. weird. `perf_event_open` has the notion of event groups, which I _think_ are started and stopped together. The general idea is to open some perf event as the "group leader" that signals if the event group should be scheduled* or not. Then, subsequent perf events are opened with a reference to the group leader to signal to the kernel that the new events are part of the group. Unfortunately, I don't see a way to get the file descriptor for an opened event (barring wildly dangerous transmutes), which we need to pass to [`set_group`](https://docs.rs/perfcnt/0.6.0/perfcnt/linux/struct.PerfCounterBuilderLinux.html#method.set_group) to wire things up right. I'll open an upstream issue, but I think it'll take a bit of tweaking in `perfcnt` to support this.

In the mean time, my hope is that at the least the variance between these counters will be constant for a given piece of hardware, so even without that fixed I think it won't be too much noise.

<details>
<summary>*why is there a scheduler involved, you may ask? I was surprised to learn this, and now you can be too:</summary>

Some x86 performance counters are constrained to only be sampled by certain performance counters, the constraints change by microarchitecture, and Linux makes a strong effort to expose a very generic and hardware-agnostic interface to the platform's performance counters. Consequently, Linux will take the active set of requested performance counters, both in the form of human-friendly names ("CPU Cycles" or "Instructions Retired" are straightforward, "{Front,Back}end Stalls"  are more complex) and raw hardware performance counter IDs, and figure out which counters can get scheduled onto which counters to provide a fair sampling of the underlying process (or system). Because performance counters can be multiplexed between "sampling an application" and "sampling the operating system" there can be some counter-juggling on every context switch.

Additionally, and this is detail we definitely don't need to care about here, Linux exposes the idea that you may only want to sample counters in a perf event group either all together, or not at all. This makes sense when trying to count rare events that you want to measure among multiple dimensions. These impose additional custom constraints that the perf event scheduler has to respect.

The whole thing is quite a bit more complicated than I'd first expected!
</details>

Reply:
I looked into this more today because when I used the code here in my sightglass fork I just had the nagging feeling that the repeated calls to `ioctl` before and after the section we want to measure were going to be too noisy. Maybe this is good enough for now but I ended up opening https://github.com/gz/rust-perfcnt/issues/19 to see if it is fixable in `perfcnt`.

For others who are interested, here are my notes:

If we want to use the `perfcnt` crate to measure events like CPU cycles we need to group the events or we risk polluting the results with kernel calls. `perfcnt` uses `perf_event_open` on Linux for measuring events; this means that most functionality available to the `perf` tool should be [available to users of `perfcnt`](https://github.com/gz/rust-perfcnt/blob/5f060f5253adfebbe91ceb8da0705ac6e651895a/src/linux/mod.rs#L631-L640). Notably, `perfcnt` is Linux-only (for now) because of the dependence on perf_event_open. `perf_event_open` promises the ability to measure multiple events simultaneously:
> A call to perf_event_open() creates a file descriptor that allows measuring performance information. Each file descriptor corresponds to one event that is measured; these can be grouped together to measure multiple events simultaneously (https://man7.org/linux/man-pages/man2/perf_event_open.2.html)

It does this with the `group_fd` parameter: the first event opened uses `group_fd = -1` and subsequent events use the first event's returned `fd` as their `group_fd`. When [`ioctl` is called to enable the counters](https://github.com/gz/rust-perfcnt/blob/5f060f5253adfebbe91ceb8da0705ac6e651895a/src/linux/mod.rs#L766-L772), all of the counters are measured together, which avoids measuring the overhead of enabling each event separately with `ioctl`.

As @iximeow mentions above, it is not always possible to have all the events we want scheduled on the same CPU:
> An event group is scheduled onto the CPU as a unit: it will be put onto the CPU only if all of the events in the group can be put onto the CPU.  This means that the values of the member events can be meaningfully compared—added, divided (to get ratios), and so on—with each other, since they have counted events for the same set of executed instructions. (https://man7.org/linux/man-pages/man2/perf_event_open.2.html)

But we can handpick events that we know will fit. 

perfcnt allows us to set the `group_fd` using [set_group](https://github.com/gz/rust-perfcnt/blob/5f060f5253adfebbe91ceb8da0705ac6e651895a/src/linux/mod.rs#L261-L264). But, as [someone smart on StackOverflow describes](https://stackoverflow.com/a/42092180), then we must be able to read the combined counter data appropriately: perfcnt can do this using [ReadRecord](https://github.com/gz/rust-perfcnt/blob/5f060f5253adfebbe91ceb8da0705ac6e651895a/src/linux/perf_format.rs#L198-L199) when combined with the [enable_read_format_id flag](https://docs.rs/perfcnt/0.6.0/perfcnt/linux/struct.PerfCounterBuilderLinux.html#method.enable_read_format_id). The resulting `Vec<<u64, Option<u64>)>` contained there can then be matched up to the event IDs to return a struct/map with the values associated correctly. Unfortunately, the parsing of the counter values has been limited to single events in the `ReadRecord` used in `src/linux/mod.rs` and the `ReadRecord` in `src/linux/perf_format.rs`, which looks to have the support, is not used: I opened an issue for this, https://github.com/gz/rust-perfcnt/issues/19.

Reply:
> I looked into this more today because when I used the code here in my sightglass fork I just had the nagging feeling that the repeated calls to ioctl before and after the section we want to measure were going to be too noisy. Maybe this is good enough for now but I ended up opening gz/rust-perfcnt#19 to see if it is fixable in perfcnt.

i'm inclined to agree that multiple `ioctl` will be noisier than necessary but also hope it's sufficient for us to get started. beyond that, thank you for your notes and opening the upstream issue!